p2p/discover/udp_quorum.go

```diff
@@ -0,0 +1,479 @@
+		gotreply:    make(chan reply),
+		addpending:  make(chan *pending),
+	}
+	udpQ := &udpQuorum{*udp}
+	tab, err := newTable(udpQ, ln.Database(), cfg.Bootnodes)
+	if err != nil {
+		return nil, nil, err
+	}
+	udpQ.tab = tab
+
+	udpQ.wg.Add(2)
+	go udpQ.loop()
+	go udpQ.readLoop(cfg.Unhandled)
+	return udpQ.tab, udpQ, nil
+}
+
+// ping sends a ping message to the given node and waits for a reply.
+func (t *udpQuorum) ping(toid enode.ID, toaddr *net.UDPAddr) error {
+	return <-t.sendPing(toid, toaddr, nil)
+}
+
+// sendPing sends a ping message to the given node and invokes the callback
+// when the reply arrives.
+func (t *udpQuorum) sendPing(toid enode.ID, toaddr *net.UDPAddr, callback func()) <-chan error {
+	req := &pingQuorum{
+		Version:    4,
+		From:       t.ourEndpoint(),
+		To:         makeEndpoint(toaddr, 0), // TODO: maybe use known TCP port from DB
+		Expiration: uint64(time.Now().Add(expiration).Unix()),
+		Hostname:   t.localNode.Node().Host(),
+	}
+	packet, hash, err := encodePacket(t.priv, pingPacket, req)
+	if err != nil {
+		errc := make(chan error, 1)
+		errc <- err
+		return errc
+	}
+	errc := t.pending(toid, pongPacket, func(p interface{}) bool {
+		ok := bytes.Equal(p.(*pongQuorum).ReplyTok, hash)
+		if ok && callback != nil {
+			callback()
+		}
+		return ok
+	})
+	t.localNode.UDPContact(toaddr)
+	t.write(toaddr, req.name(), packet)
+	return errc
+}
+
+// findnode sends a findnode request to the given node and waits until
+// the node has sent up to k neighbors.
+func (t *udpQuorum) findnode(toid enode.ID, toaddr *net.UDPAddr, target encPubkey) ([]*node, error) {
+	// If we haven't seen a ping from the destination node for a while, it won't remember
+	// our endpoint proof and reject findnode. Solicit a ping first.
+	if time.Since(t.db.LastPingReceived(toid)) > bondExpiration {
+		t.ping(toid, toaddr)
+		t.waitping(toid)
+	}
+
+	nodes := make([]*node, 0, bucketSize)
+	errc := t.pending(toid, neighborsPacket, func(r interface{}) bool {
+		reply := r.(*neighborsQuorum)
+		for _, rn := range reply.Nodes {
+			n, err := t.nodeFromRPC(toaddr, rn)
+			if err != nil {
+				log.Trace("Invalid neighbor node received", "ip", rn.IP, "addr", toaddr, "err", err)
+				continue
+			}
+			nodes = append(nodes, n)
+		}
+		return len(reply.Nodes) >= bucketSize
+	})
+	t.send(toaddr, findnodePacket, &findnodeQuorum{
+		Target:     target,
+		Expiration: uint64(time.Now().Add(expiration).Unix()),
+		IsQuorum:   true,
+	})
+	return nodes, <-errc
+}
+
+func (t *udpQuorum) handleReply(from enode.ID, ptype byte, req packetQuorum) bool {
+	matched := make(chan bool, 1)
+	select {
+	case t.gotreply <- reply{from, ptype, req, matched}:
+		// loop will handle it
+		return <-matched
+	case <-t.closing:
+		return false
+	}
+}
+
+var (
+	// Neighbors replies are sent across multiple packets to
+	// stay below the 1280 byte limit. We compute the maximum number
+	// of entries by stuffing a packet until it grows too large.
+	maxNeighborsQuorum = 3
```

Comment:
this value is calculated in `udp.go#init()`, why do we hard code here?

Reply:
The value calculated in `udp.go#init()` (12, if I remember correctly) is so because a record in discovery V4 is a fixed size, so you can always fit 12 records in a single packet. 
Since hostnames are a variable size, this is no longer true. 3 was fairly arbitrary but unlikely to ever reach the limit.
I'll change it to check the size after each node has been added and see if it has gone over the limit.